src/pytorch_lightning/plugins/environments/azure_openmpi_environment.py

```diff
@@ -0,0 +1,77 @@
+# Copyright The PyTorch Lightning team.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from pytorch_lightning.plugins.environments.cluster_environment import ClusterEnvironment
+
+
+class AzureOpenMPIEnvironment(ClusterEnvironment):
+    """Environment for an OpenMPI environment on Azure.
+
+    See Azure documentation here: https://docs.microsoft.com/en-us/azure/machine-learning/how-to-train-distributed-gpu#mpi
+    """
+
+    def __init__(self, devices: int = 1) -> None:
```

Comment:
I would prefer not pass devices in here. Do you think we could define the node_rank in a different way?
If I look at the docs [here](https://www.open-mpi.org/faq/?category=running#mpi-environmental-variables), I can see env variables for node rank. Is this also getting set on azure?

Reply:
Unfortunately, the `OMPI_COMM_WORLD_NODE_RANK` variable doesn't appear to correspond to node rank. You can see the note about this in the [Azure documentation](https://docs.microsoft.com/en-us/azure/machine-learning/how-to-train-distributed-gpu#environment-variables-from-open-mpi):

> Despite the name, environment variable OMPI_COMM_WORLD_NODE_RANK does not corresponds to the NODE_RANK. To use per-node-launcher, set process_count_per_node=1 and use OMPI_COMM_WORLD_RANK as the NODE_RANK.

The Open MPI docs also define this variable as:

> OMPI_COMM_WORLD_NODE_RANK - the relative rank of this process on this node looking across ALL jobs.

I just ran a job on a 2 node setup with 2 V100 GPUs per node using Microsoft's docker image `mcr.microsoft.com/azureml/openmpi4.1.0-cuda11.3-cudnn8-ubuntu20.04` and it is setting `OMPI_COMM_WORLD_NODE_RANK` and `OMPI_COMM_WORLD_LOCAL_RANK` to the same values:

Driver 0
```sh
OMPI_COMM_WORLD_RANK = 0
OMPI_COMM_WORLD_LOCAL_RANK = 0
OMPI_COMM_WORLD_SIZE = 4
OMPI_COMM_WORLD_LOCAL_SIZE = 2
OMPI_COMM_WORLD_NODE_RANK = 0
OMPI_UNIVERSE_SIZE = 4
```

Driver 1
```sh
OMPI_COMM_WORLD_RANK = 1
OMPI_COMM_WORLD_LOCAL_RANK = 1
OMPI_COMM_WORLD_SIZE = 4
OMPI_COMM_WORLD_LOCAL_SIZE = 2
OMPI_COMM_WORLD_NODE_RANK = 1
OMPI_UNIVERSE_SIZE = 4
```

Driver 2
```sh
OMPI_COMM_WORLD_RANK = 2
OMPI_COMM_WORLD_LOCAL_RANK = 0
OMPI_COMM_WORLD_SIZE = 4
OMPI_COMM_WORLD_LOCAL_SIZE = 2
OMPI_COMM_WORLD_NODE_RANK = 0
OMPI_UNIVERSE_SIZE = 4
```

Driver 3
```sh
OMPI_COMM_WORLD_RANK = 3
OMPI_COMM_WORLD_LOCAL_RANK = 1
OMPI_COMM_WORLD_SIZE = 4
OMPI_COMM_WORLD_LOCAL_SIZE = 2
OMPI_COMM_WORLD_NODE_RANK = 1
OMPI_UNIVERSE_SIZE = 4
```